modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -484,6 +537,189 @@ public int getZoneId(String name) {
+                        zonesConfiguration.distributionZones();
+
+                for (int i = 0; i < zones.value().size(); i++) {
+                    DistributionZoneView zone = zones.value().get(i);
+
+                    if (zone.zoneId() == zoneId) {
+                        immediateScaleUp0 = zone.dataNodesAutoAdjustScaleUp() == 0;
+                        immediateScaleDown0 = zone.dataNodesAutoAdjustScaleDown() == 0;
+
+                        break;
+                    }
+                }
+            }
+
+            immediateScaleUp.set(immediateScaleUp0);
+            immediateScaleDown.set(immediateScaleDown0);
+        });
+
+        CompletableFuture<Void> topVerScaleUpFut = new CompletableFuture<>();
+
+        topVerFut.thenAccept(ignored -> {
+            synchronized (dataNodesMutex) {
+                ZoneState zoneState = zonesState.get(zoneId);
+
+                if (zoneState == null) {
+                    throw new DistributionZoneNotFoundException(zoneId);
+                }
+
+                CompletableFuture<Void> topVerScaleUpFut0 = null;
+
+                if (immediateScaleUp.get()) {
+                    Map.Entry<Long, Long> scaleUpRevisionEntry = topVerAndScaleUpRevision.ceilingEntry(topVer);
+
+                    Long scaleUpRevision = null;
+
+                    if (scaleUpRevisionEntry != null) {
+                        scaleUpRevision = scaleUpRevisionEntry.getValue();
+                    }
+
+                    if (scaleUpRevision != null && zoneState.scaleUpRevision() < scaleUpRevision) {
+                        Map.Entry<Long, CompletableFuture<Void>> ceilingEntry =
+                                zoneState.revisionScaleUpFutures().ceilingEntry(scaleUpRevision);
+
+                        if (ceilingEntry != null) {
+                            topVerScaleUpFut0 = ceilingEntry.getValue();
+                        }
+
+                        if (topVerScaleUpFut0 == null) {
+                            topVerScaleUpFut0 = new CompletableFuture<>();
+
+                            zoneState.revisionScaleUpFutures().put(scaleUpRevision, topVerScaleUpFut0);
+                        }
+
+                        topVerScaleUpFut0.handle((ignored0, e) -> {
+                            if (e == null) {
+                                topVerScaleUpFut.complete(null);
+                            } else {
+                                topVerScaleUpFut.completeExceptionally(e);
+                            }
+
+                            return null;
+                        });
+                    } else {
+                        topVerScaleUpFut.complete(null);
+                    }
+                } else {
+                    topVerScaleUpFut.complete(null);
+                }
+            }
+        });
+
+        CompletableFuture<Void> topVerScaleDownFut = new CompletableFuture<>();
+
+        topVerFut.thenAccept(ignored -> {
```

Comment:
The chaining is incorrect. 651::topVerFut.thenAccept will be executed before 598::topVerFut.thenAccept, so that immediateScaleDown will have uninitialized value (false). By the way, why there are no failing tests showing that?

Reply:
Actually the chaining is correct. I updated it to make it more clear.